snuba/subscriptions/verifier.py

```diff
@@ -99,3 +112,203 @@ def close(self, timeout: Optional[float] = None) -> None:
+        self.__timestamp_low_watermark: Optional[int] = None
+
+        # The most recent timestamp we've seen
+        self.__timestamp_high_watermark: Optional[int] = None
+
+        self.__threshold_sec = threshold_sec
+        self.__metrics = metrics
+
+    def increment(self, topic: ResultTopic, item: SubscriptionResultData) -> None:
+        # When we get the first message, set the low watermark
+        if self.__timestamp_low_watermark is None:
+            self.__timestamp_low_watermark = item.timestamp + self.__threshold_sec
+
+        # Advance the high watermark if needed
+        if (
+            self.__timestamp_high_watermark is None
+            or item.timestamp > self.__timestamp_high_watermark
+        ):
+            self.__timestamp_high_watermark = item.timestamp
+
+        # Ensure the message is not below the low watermark
+        if item.timestamp <= self.__timestamp_low_watermark:
+            # Record `stale_message` unless the consumer is just starting up
+            if (
+                self.__timestamp_high_watermark
+                >= self.__timestamp_low_watermark + self.__threshold_sec
+            ):
+                self.__metrics.increment("stale_message")
+            return
+
+        # Increment counters
+        store = self.__stores[topic]
+
+        if item.timestamp not in store:
+            store[item.timestamp] = 0
+        store[item.timestamp] += 1
+
+        # Record metrics and advance the low watermark
+        low_watermark = self.__timestamp_high_watermark - self.__threshold_sec
+
+        if low_watermark > self.__timestamp_low_watermark:
+            self._record(low_watermark)
+            self.__timestamp_low_watermark = low_watermark
+
+    def _record(self, until: int) -> None:
+        """
+        Records metrics up to and including `until` then drop recorded elements.
+        """
+        while True:
+            next_orig_ts = next(iter(self.__stores[ResultTopic.ORIGINAL]), None)
+            next_new_ts = next(iter(self.__stores[ResultTopic.NEW]), None)
+
+            if next_orig_ts is not None and next_orig_ts > until:
+                next_orig_ts = None
+            if next_new_ts is not None and next_new_ts > until:
+                next_new_ts = None
+
+            # Set next_ts
+            if next_orig_ts is None and next_new_ts is None:
+                next_ts = None
+            elif next_orig_ts is None:
+                next_ts = next_new_ts
+            elif next_new_ts is None:
+                next_ts = next_orig_ts
+            else:
+                next_ts = min([next_orig_ts, next_new_ts])
+
+            # Nothing else to remove
+            if next_ts is None:
+                return
+
+            orig_count = self.__stores[ResultTopic.ORIGINAL].pop(next_ts, 0)
+            new_count = self.__stores[ResultTopic.NEW].pop(next_ts, 0)
+            self.__metrics.increment("result_count_diff", new_count - orig_count)
```

Comment:
For this metric, is it correct to say that lower absolute value is better but we probably won't hit 0 that often?

Or should this be 0 most of the time with occasional non-zero values when things can't be kept in sync?

Reply:
Yeah, that's correct. The verifier does make an effort to not log anything when things are not quite in sync and wait 30 seconds in general for things to get in sync so we are definitely aiming for 0s all the time.

Reply:
I was thinking about this, and the value should be wrapped in an `abs`. Otherwise you could report both a +1 and -1 at basically the same time, and Datadog will add them together.

Reply:
Yeah, good point. I forgot Datadog adds these together in each interval before sending. Updated.